### PR TITLE
Add initial BGS client and RabbitMQ consumer

### PR DESCRIPTION
**Pull request by nanotone**
Initial implementation of a microservice that consumes RabbitMQ messages and makes BGS API calls 🎉 

Ticket: #1289
[Slack thread](https://dsva.slack.com/archives/C04QLHM9LR0/p1680023668433189) where @yoomlam and I hammered out the JSON payload format

## How to use this microservice
- Post JSON-formatted on the `bgs-api` exchange, to the `add-note` queue.
- Request type 1 -- adding multiple claim-level notes:
```json
{"vbmsClaimId": "1234", "claimNotes": ["First note", "Second note"]}
```

- Request type 2 -- adding a single veteran-level note:
```json
{"vbmsClaimId": "1234", "veteranNote": "A veteran-level note"}
# or, supply the veteran's participant ID directly
{"veteranParticipantId": "2468", "veteranNote": "A veteran-level note"}
```
If the veteran's participant ID is omitted, it will be discovered as the veteran associated with the claim ID. If the veteran's participant ID is provided, any claim ID also provided will be ignored.

## Example log
```
D, [2023-04-04T20:08:11.651370 #1] DEBUG -- : delivery_info: {:consumer_tag=>"bunny-1680638887000-323701686295", :delivery_tag=>#<Bunny::VersionedDeliveryTag:0x00007f29f000fdb8 @tag=1, @version=0>, :redelivered=>false, :exchange=>"bgs-api", :routing_key=>"add-note", :consumer=>#<Bunny::Consumer:2120 @channel_id=1 @queue=add-note @consumer_tag=bunny-1680638887000-323701686295 @exclusive=false @no_ack=true>, :channel=>#<Bunny::Channel:2140 @id=1 @connection=#<Bunny::Session:0x1040 guest@rabbitmq-service:5672, vhost=/, addresses=[rabbitmq-service:5672]> @open=true>}
Subscriber received request {"claimNotes"=>["a note"]}
I, [2023-04-04T20:08:11.651654 #1]  INFO -- : Response: {:statusCode=>400, :statusMessage=>"ArgumentError: vbmsClaimId is required for claimNotes"}

D, [2023-04-04T20:08:50.320279 #1] DEBUG -- : delivery_info: {:consumer_tag=>"bunny-1680638887000-323701686295", :delivery_tag=>#<Bunny::VersionedDeliveryTag:0x00007f29f000c938 @tag=2, @version=0>, :redelivered=>false, :exchange=>"bgs-api", :routing_key=>"add-note", :consumer=>#<Bunny::Consumer:2120 @channel_id=1 @queue=add-note @consumer_tag=bunny-1680638887000-323701686295 @exclusive=false @no_ack=true>, :channel=>#<Bunny::Channel:2140 @id=1 @connection=#<Bunny::Session:0x1040 guest@rabbitmq-service:5672, vhost=/, addresses=[rabbitmq-service:5672]> @open=true>}
Subscriber received request {"vbmsClaimId"=>123, "claimNotes"=>["a note"]}
D, [2023-04-04T20:08:50.321620 #1] DEBUG -- : HTTPI /none GET request to beplinktest.vba.va.gov (httpclient)
I, [2023-04-04T20:08:50.449938 #1]  INFO -- : Response: {:statusCode=>500, :statusMessage=>"SocketError: getaddrinfo: Name or service not known (beplinktest.vba.va.gov:80)"}
```

As expected, BGS LinkTest is not accessible from my local Docker container.

## Further work
- ~I've only updated the Dockerfile to use the stabilized Gemfile.lock, but haven't addressed this [issue with ruby-dev version mismatches](https://dsva.slack.com/archives/C04QLHM9LR0/p1680317638263399); locally, I just use the non-alpine base image and comment out all the APK commands.~
  - Addressed in 8cbfb7f.
- ~Higher BGS environments (preprod, prod) require TLS. The straightforward way to configure this is to update `config/settings.yml` with higher environments containing config values for the `ssl_` keys (see `config/bgs_initializer.rb`) but these require the cert/key/cacert to live on the filesystem. Not sure how this will play out with our deployment process, as it's quite different from how the [BIP client microservice loads credentials from b64-encoded env vars](https://github.com/department-of-veterans-affairs/abd-vro/wiki/BIP-APIs#java-implementation-and-code-walkthrough).~
  - Addressed in a handful of commits in this PR, particularly e0a20fa. The `entrypoint.sh` now looks for the same env vars as the BIP microservice, and creates cert/key/cacert files relative to `main_consumer.rb`.
- ~Currently, the BGS gem constructs the API's URL from the `env` config variable, which won't work with Mock BGS. `config/bgs_initializer.rb` needs to be updated to support supplying the entire hostname. And it'll likely be an env var, rather than a key in `settings.yml`.~
  - The process now accepts a `$BGS_BASE_URL` (in the form `http://mock-bgs-api:8080`) that points the BGS client at an alternate API server -- i.e. a forward proxy, jumpbox, or mock server.

### CODE REVIEW

**Comment by yoomlam** [comment]
> update config/settings.yml with higher environments containing config values for the ssl_ keys (see config/bgs_initializer.rb) but these require the cert/key/cacert to live on the filesystem.

We can mount secrets as files on the filesystem. Are [these](https://github.com/department-of-veterans-affairs/abd-vro/pull/1417/files#diff-6ff915f2eb0353c21e671cbf0f908d3833cebcc237a7e80709c7ba06fd5a08daR17-R19) the secrets?

**Comment by nanotone** [comment]
@yoomlam Yep. In manual testing, I supplied `tls_bip.crt`, passwordless `tls.key` (both unpacked from `$BIP_TRUSTSTORE`), and `cacert.pem` (from `$BIP_TRUSTSTORE`) to these three config values.

I haven't looked at secrets in Kubernetes, though I assume that would've made life easier.

**Comment by yoomlam** [comment]
> @yoomlam Yep. In manual testing, I supplied `tls_bip.crt`, passwordless `tls.key` (both unpacked from `$BIP_TRUSTSTORE`), and `cacert.pem` (from `$BIP_TRUSTSTORE`) to these three config values.
> 
> I haven't looked at secrets in Kubernetes, though I assume that would've made life easier.

@nanotone What are the commands to create those files from the `$BIP_TRUSTSTORE`?

**Comment by nanotone** [comment]
>What are the commands to create those files from the $BIP_TRUSTSTORE?

First, convert the env var from b64-with-spaces to binary and write to filesystem. I know you can do it with command-line tools, but it's easier and more readable in Ruby:
```ruby
require "base64"
File.write("keystore.p12", Base64.decode64(BIP_KEYSTORE.gsub(/\s/, "")))
File.write("truststore.p12", Base64.decode64(BIP_TRUSTSTORE.gsub(/\s/, "")))
```

Then they need to be PKCS12-decrypted:
```bash
openssl pkcs12 -in keystore.p12 -password pass:$BIP_PASSWORD -out tls_bip.crt -nokeys
openssl pkcs12 -in keystore.p12 -password pass:$BIP_PASSWORD -out tls.key -nocerts -nodes
openssl pkcs12 -in truststore.p12 -password pass:$BIP_PASSWORD -out va_all.crt
# now we can verify that it works
curl -v --cert tls_bip.crt --key tls.key --cacert va_all.crt https://bepprep.vba.va.gov
```

File: app/src/docker/docker-compose.yml [hunk 1]
```diff
@@ -356,6 +356,8 @@ services:
     environment:
       <<: *common-vars
       <<: *rabbitmq-placeholder-vars
+      # A mock-bgs-api needs to be created and exposes an unused port
+      BGS_BASE_URL: http://mock-bgs-api:9020
```
Comment (yoomlam) [inline app/src/docker/docker-compose.yml:360]: Expecting @greene858 to provide the mock-bgs-api container. Note other mock containers also expose ports, so don't collide with them.

File: svc-bgs-api/src/docker/Dockerfile [hunk 2]
```diff
@@ -1,10 +1,11 @@
-FROM ruby:2.7.7-alpine3.16
- 
+#FROM ruby:2.7.7-alpine3.16
+FROM ruby:2.7.7
```
Comment (yoomlam) [inline svc-bgs-api/src/docker/Dockerfile:2]: Ran into a [build error](https://github.com/department-of-veterans-affairs/abd-vro-internal/actions/runs/4629816819) and had to make this change and comment out `apk` commands below. 
@greene858, this blocks deployment and testing unless you resolve SecRel alerts from not using the `alpine` image.

File: app/src/docker/docker-compose.yml (hunk 1 above)
Comment (nanotone) [reply app/src/docker/docker-compose.yml:360]: Draft PR for mock-bgs-api is open: https://github.com/department-of-veterans-affairs/abd-vro/pull/1420

File: svc-bgs-api/src/config/bgs_initializer.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  # $BGS_BASE_URL also determines whether HTTPS is used.
+  config.jumpbox_url = ENV["BGS_BASE_URL"]
+  config.env = nil
+  if config.jumpbox_url.downcase.start_with?("https:")
+    config.ssl_cert_file = settings["ssl_cert_file"]
+    config.ssl_cert_key_file = settings["ssl_cert_key_file"]
+    config.ssl_ca_cert = settings["ssl_ca_cert_file"]
+    config.ssl_verify_mode = settings["ssl_verify_mode"] || "none"
+  else
+    config.ssl_verify_mode = "none"
```
Comment (yoomlam) [inline svc-bgs-api/src/config/bgs_initializer.rb:29]: Trivial suggestion that reduces LOC.
```suggestion
  config.ssl_verify_mode = "none"
  if config.jumpbox_url.downcase.start_with?("https:")
    config.ssl_cert_file = settings["ssl_cert_file"]
    config.ssl_cert_key_file = settings["ssl_cert_key_file"]
    config.ssl_ca_cert = settings["ssl_ca_cert_file"]
    config.ssl_verify_mode = settings["ssl_verify_mode"]
```

File: svc-bgs-api/src/config/settings.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Settings (like the RabbitMQ connection options or API keys)
+# Loaded by setup.rb and put into a global variable.
+
+bgs:
+  application: VRO
+  client_station_id: 281
+  client_username: VROSYSACCT
+  ssl_cert_file: /app/certs/tls_bip.crt
+  ssl_cert_key_file: /app/certs/tls.key
+  ssl_ca_cert_file: /app/certs/va_all.crt
```
Comment (yoomlam) [inline svc-bgs-api/src/config/settings.yml:10]: So for local testing, these `ssl` settings shouldn't be needed. Cool!

File: svc-bgs-api/src/config/setup.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  $fdio
+end
+
+if(ENV['DOCKER_LOGS'] == '1' && File.exist?('/proc/1/fd/1'))
+  redirect_output("| ")
+end
+
+puts "Running setup.rb"
+
+
```
Comment (yoomlam) [inline svc-bgs-api/src/config/setup.rb:24]: I'd expect a linter to remove 2 consecutive blank lines. Can we add a Ruby linter, maybe as a follow-on ticket?

File: svc-bgs-api/src/config/settings.yml (hunk 4 above)
Comment (nanotone) [reply svc-bgs-api/src/config/settings.yml:10]: Yep. I figured if we wanted to make settings/config more consistent and constant across environments, it's easiest to handle moving parts in code.

File: svc-bgs-api/src/lib/rabbit_subscriber.rb [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    attempts ||= 1
+    connection.start
+  rescue => error
+    $logger.error error.message
+    if (attempts += 1) < 5
+      $logger.warn "Retrying to connect to RabbitMQ in 10 seconds"
+      sleep 10
+      retry
+    else
+      raise error
```
Comment (yoomlam) [inline svc-bgs-api/src/lib/rabbit_subscriber.rb:25]: What happens as a result of this `raise`? Will ruby exit?

**Review by yoomlam — APPROVED** [review]
Remember to add unit tests.

File: svc-bgs-api/src/lib/rabbit_subscriber.rb (hunk 6 above)
Comment (nanotone) [reply svc-bgs-api/src/lib/rabbit_subscriber.rb:25]: Yes, the process will exit in 40 seconds if all attempts to connect to RabbitMQ fail.

**Review by yoomlam — APPROVED** [review]

File: svc-bgs-api/src/docker/Dockerfile (hunk 2 above)
Comment (greene858) [reply svc-bgs-api/src/docker/Dockerfile]: I revised the app and Dockerfile to address this

**Review by greene858 — APPROVED** [review]

**Pull request merged**